### DISCUSSION TITLE
ide integration / language server

### DISCUSSION THREAD

**Discussion by DetachHead** [KotlinIsland/basedmypy#347] [Ideas] [answered]
VSCode's Python LSP (pyright) and PyCharm's type analysis don't infer/support some of the features that basedmypy does, so IDE features like auto complete don't work in all cases. This will be a growing issue as more wacky features are added.

There is an upstream issue for integrating an LSP which would resolve these issues (although not so easilly for PyCharm, which would require a custom plugin that would proxy to the LSP).

### Original post
```py
@overload
def foo(value: int): ...

@overload
def foo(value: str): ...

def foo(value): # inferred as int | str
  value # no ide autocomplete
```

the more features basedmypy gets the bigger of a problem this will become. we need a way for IDEs to support based features

some ideas:

- language server (sounds super hard)
- port basedmypy features to pyright (basedpyright?) - this will solve the problem for vscode but not pycharm

**Comment by KotlinIsland** [comment]
This sounds like a massive work item

**Comment by Zeckie** [comment]
> This sounds like a massive work item

Or a whole separate project (eg if it is writing basedmypy aware plugin/s). There are lots of other IDE integration features that would be useful, like showing errors - https://github.com/KotlinIsland/basedmypy/discussions/134#discussioncomment-3119029.

**Reply by KotlinIsland** [reply]
> like showing errors

There's no reason that basedmypy would be incompatible with existing IDE integrations.

> Or a whole separate project

LSP integration would likely be part of mypy/basedmypy, although IntelliJ support would need a wrapper plugin, which would likely require a separate project.

**Comment by DetachHead** [comment, marked as answer]
> port basedmypy features to pyright (basedpyright?)

this is the solution i went with: https://github.com/detachhead/basedpyright

> this will solve the problem for vscode but not pycharm

intellij now supports language servers so we can probably make a pycharm plugin now https://github.com/DetachHead/basedpyright/issues/27

**Reply by KotlinIsland** [reply]
Additionally, there is the basedtyping plugin for IntelliJ.

A fully fledged mypy language integration would be far too much work.
